# History

## 1.0.2 (2024-04-27)

This patch release fixes an issue with not capturing an EmptyDataError when reading VCF files and an issue where Clair3 
VCFs from non nf-flu analyses are prematurely converted to a dict from a Pandas DataFrame.

### Changes

- dev: Update type hints given mypy and ruff linting
- fix: check BAM files for reference sequence ID/name in addition to Mosdepth files. Improved searching for the reference sequence GFF and FASTA in the input directory before trying to fetch from NCBI. 

## 1.0.1 (2024-02-20)

Fix issue with `model_dump` method not being available in Pydantic V1 (#55) so this patch release makes it so that Pydantic V2 is required.

## 1.0.0 (2024-02-13)

This is a major release of wgscovplot with many user experience and quality of life changes to make the tool easier to
run and use. The frontend (HTML output) has been almost completely rewritten to have more of a single-page app (SPA) 
like feel.

There are also a number of improvements to the developer experience so that frontend dev can be done separately with 
live-code reloading using real-world datasets.

- Calculate depths from BAM files if Mosdepth `per-base.bed.gz` not found.
- SolidJS instead of JQuery for UI
- Migrated to Typescript and TSX from JS for frontend code.
- Hatch for packaging, building and managing project. Migrated from `setup.py` to `pyproject.toml`.
- Removed compiled JS from repo. JS/TS is compiled at build time with npm or bun if installed (`hatch_build.py`).
- Migrated from Webpack to Vite for frontend dev and JS compilation.
- Migrated from Bootstrap CSS to Tailwind CSS
- Added frontend test datasets and profiles for isolated frontend dev. See `web/README.md` for details.
- Moved CLI test data for easier testing with segmented and non-segmented virus pipeline output
- Added Black formatting, Ruff linting and Mypy type checking.
- Fixed variant bar colours to represent ALT allele rather than REF.
- Simplified CLI options with auto-detection of files and parameters based on type of analysis results provided.
- Coverage depths are compressed by default to base64 encoded string in HTML output to reduce HTML output size.
- Added more coverage plot visual customization options for colours, fonts, sizes, etc.
- Removed Jest JS/TS testing for now.

## 0.3.0 (2022-08-31)

- Add support for segment virus (AIV)
- Accept result folder generated by `nf-flu` with option `--segment-virus`
- Annotate primer sequences with option `--primer-seq`. The primer sequence in fasta file

## 0.2.1 (2022-06-14)

- Fixed issues of tooltips displayed incorrectly due to samples list being reversed 

## 0.2.0 (2022-03-28)

- Automatically Gene Bank and reference retrieval
- Automatically detect amplicon data and option to toggle amplicon bar
- Variant heat map is updated according to selected samples
- Downgrade and use Echarts v5.2.2 to fix the problem of amplicon plot (https://github.com/nhhaidee/wgscovplot/issues/36)
- Clean up code.

## 0.1.1 (2022-03-18)

Fixes parsing of Medaka VCF with low/no coverage variants.

## 0.1.0 (2022-03-04)

Initial alpha release.
